Hud10837/find route kotlin

hud10837 opened the pull request:
Hello! This one was fun--the actual logic of the sample didn't really change from the java version, but creating a bottom sheet to replace the navigation was a bit of a challenge. I hope it's ended up cleaner, but please give me any structural suggestions you like. @Rachael-E could you do a first review?

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+      mapView.map = map
+    }
+
+    // ensure the floating action button moves to be above the attribution view
+    val params = directionFab.layoutParams as ConstraintLayout.LayoutParams
+    mapView.addAttributionViewLayoutChangeListener { _, _, _, _, bottom, _, _, _, oldBottom ->
+      val heightDelta = bottom - oldBottom
+      params.bottomMargin += heightDelta
+    }
+
+    // hide the bottom sheet
+    bottomSheetBehavior.peekHeight = 0
+    // animate the arrow when the bottom sheet slides
+    bottomSheetBehavior.addBottomSheetCallback(object : BottomSheetCallback() {
+      override fun onSlide(bottomSheet: View, slideOffset: Float) {
+        bottomSheet.header.imageView.rotation = slideOffset * 180f
+      }
+
+      override fun onStateChanged(bottomSheet: View, newState: Int) {
+        bottomSheet.header.imageView.rotation = when (newState) {
+          BottomSheetBehavior.STATE_EXPANDED -> 180f
+          else -> 0f
+        }
+      }
+    })
+    // while the bottom sheet is hidden, make the map view span the whole screen
+    val mainContainerParams = mainContainer.layoutParams as CoordinatorLayout.LayoutParams
+    mainContainerParams.bottomMargin = 0
```
I think a lot of this really neat UI should be hidden away in a method where possible: we want the main code focus to be on our API code which should be first and foremost in the sample. If we can tuck this UI logic away into a method and just call that method here, that will be a lot more tidy and allow customers to hone in on our API code more easily :)

hud10837 replied:
You are 100% right! Will address in my pr comment

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+            progressDialog.dismiss()
+          }
+
+          // expand or collapse the bottom sheet when the header is clicked
+          bottomSheet.header.setOnClickListener {
+            bottomSheetBehavior.state = when (bottomSheetBehavior.state) {
+              BottomSheetBehavior.STATE_COLLAPSED -> BottomSheetBehavior.STATE_EXPANDED
+              else -> BottomSheetBehavior.STATE_COLLAPSED
+            }
+          }
+
+          bottomSheet.directionsListView.apply {
+            // Set the adapter for the list view
+            adapter = ArrayAdapter(
+              this@MainActivity,
+              android.R.layout.simple_list_item_1,
+              directionsArray
+            )
```
I think we should hide this in a method to keep the focus on our API code. `bottomSheetHandler` or something like that if it's possible to extract it?

hud10837 replied:
see above

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+  }
+
+  /**
+   * Set up the source, destination and route symbols
```
```suggestion
   * Set up the source, destination and route symbols.
```

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+   * Set up the source, destination and route symbols
+   */
+  private fun setupSymbols() {
+    //[DocRef: Name=Picture Marker Symbol Drawable-android, Category=Fundamentals, Topic=Symbols and Renderers]
```
I would check with @TADraeseke but I don't think we need this doc style for the Kotlin samples, but check with Trevor for confirmation.

hud10837 replied:
I checked with Trevor and Mark, and Mark said that these couldn't be removed. But you make a fine point, perhaps it's unneeded on kotlin?

hud10837 replied:
Just checked with mark again and you are right, it doesn't need to be in kotlin samples!

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+   */
+  private fun setupSymbols() {
+    //[DocRef: Name=Picture Marker Symbol Drawable-android, Category=Fundamentals, Topic=Symbols and Renderers]
+    // Create a picture marker symbol from an app resource
```
```suggestion
    // create a picture marker symbol from an app resource
```

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+  /**
+   * Set up the source, destination and route symbols
+   */
+  private fun setupSymbols() {
```
I noticed that on my device the pin symbols look huge! They could do with a 50% reduction in size.

hud10837 replied:
Hopefully it's not too small now! I changed it to 30 width and height

hud10837 replied:
(was 40 I think)

Rachael-E commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+      Log.e(TAG, message)
+      Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+    }
+    //[DocRef: END]
```
this and other occurrences of it can be removed assuming we're not using doc refs in the kotlin samples

Rachael-E commented on kotlin/find-route/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,44 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:layout_marginBottom="64dp">
+
```
can remove extra line

Rachael-E commented on kotlin/find-route/src/main/res/layout/bottom_sheet.xml:
```diff
@@ -0,0 +1,45 @@
+    android:layout_height="wrap_content"
+    android:orientation="vertical">
+
+
```
can remove extra line

hud10837 commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,321 @@
+      ArcGISVectorTiledLayer(resources.getString(R.string.navigation_vector))
+    )
+    // create a map with the basemap
+    ArcGISMap(basemap).let { map ->
```
You're right, I've been trying to be more consistent in recent samples. Thanks for catching this!

hud10837 commented:
Thanks for that @Rachael-E ! Your comments ended up with me doing a lot of reworking, which was really helpful. All of the bottom sheet logic is in its own function, `setupBottomSheet()`, including the creation of the bottom sheet behavior; this means that the only logic in onCreate() that refers to the bottom sheet is just two lines to make it invisible at the start. I hope this looks a lot cleaner now, though I see now that the bottom sheet added ~70 lines of code and I hope that's not too much.

Could you take another look when you get the chance?

TADraeseke commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,325 @@
+   */
+  private fun setupSymbols() {
+    try {
+      val startDrawable =
+        ContextCompat.getDrawable(this, R.drawable.ic_source) as BitmapDrawable?
+
+      val pinSourceSymbol = PictureMarkerSymbol.createAsync(startDrawable).get()
+      pinSourceSymbol.apply {
+        loadAsync()
+        addDoneLoadingListener {
+          // add a new graphic as start point
+          val sourcePoint = Point(
+            -117.15083257944445,
+            32.741123367963446,
+            SpatialReferences.getWgs84()
+          )
+          // add it to the graphics overlay
+          graphicsOverlay.graphics.add(Graphic(sourcePoint, this))
+        }
+        // make the graphic smaller
+        width = 30f
+        height = 30f
+        offsetY = 20f
+      }
```
```suggestion
(ContextCompat.getDrawable(this, R.drawable.ic_source) as? BitmapDrawable)?.let {
        val pinSourceSymbol = PictureMarkerSymbol.createAsync(it).get().apply {
          // make the graphic smaller
          width = 30f
          height = 30f
          offsetY = 20f
        }
        // create a point for the new graphic
        val sourcePoint = Point(
          -117.15083257944445, 32.741123367963446, SpatialReferences.getWgs84()
        )
        // add it to the graphics overlay
        graphicsOverlay.graphics.add(Graphic(sourcePoint, pinSourceSymbol))
      }
```

This is maybe closer to our builder pattern? Happy to hear your thoughts on this!

hud10837 replied:
I'm not sure. I feel like because getting the drawable is a fairly complicated line, and this kind of obscures its creation. If we were to use the let{} block, I could perhaps see something like
```
val startDrawable =
        ContextCompat.getDrawable(this, R.drawable.ic_source) as BitmapDrawable?
startDrawable.let{...}
```
But then it's not really doing much, because only the picture marker symbol requires the drawable.

You removed the loadAsync/addDoneLoadingListener inside the picturemarkersymbol.apply block. Are those not necessary because we already called .get()?

TADraeseke commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,325 @@
+      Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+    }
+    try {
+      val endDrawable =
+        ContextCompat.getDrawable(this, R.drawable.ic_destination) as BitmapDrawable?
+
+      val pinDestinationSymbol = PictureMarkerSymbol.createAsync(endDrawable).get()
+      pinDestinationSymbol.apply {
+        loadAsync()
+        addDoneLoadingListener {
+          // add a new graphic as end point
+          val destinationPoint = Point(
+            -117.15557279683529,
+            32.703360305883045,
+            SpatialReferences.getWgs84()
+          )
+          // add it to the graphics overlay
+          graphicsOverlay.graphics.add(Graphic(destinationPoint, this))
+        }
+        // make the graphic smaller
+        width = 30f
+        height = 30f
+        offsetY = 20f
+      }
```
maybe replicate the above pattern?

TADraeseke commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,325 @@
+    }
+  }
+
+  /** Create a progress dialog box for tracking the route task.
+   *
+   * @param routeTask the route task progress to be tracked
+   * @return an AlertDialog set with the dialog layout view
+   */
+  private fun createProgressDialog(routeTask: RouteTask) =
+    AlertDialog.Builder(this@MainActivity)
+      .setTitle("Solving route...")
+      // provide a cancel button on the dialog
+      .setNegativeButton("Cancel") { _, _ ->
+        routeTask.cancelLoad()
+      }
+      .setCancelable(false)
+      .setView(R.layout.dialog_layout)
+      .create()
```
I don't think the calculation warrant it's own progress dialog. Maybe just overlay an invisible indeterminate progress indicator in the middle of the map view and toggle it's visibility based on a done loading listener

hud10837 replied:
This is a good idea!

TADraeseke commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,325 @@
+   *
+   * @param directions a list of DirectionManeuver which represents the route
+   */
+  private fun setupBottomSheet(directions: List<DirectionManeuver>) {
```
alot of the calls in this function look like they could be conflated into

bottomSheetBehaviour.apply {
...
...
...
}

and 

bottomSheet.apply {
...
}

hud10837 replied:
I've made this change but it created a nested apply{apply{}} block. Are we trying to avoid those? I can take it out if needed.

TADraeseke replied:
I think the directionsListView.apply can be moved out of the bottomSheetBehavious apply block, but maybe that's even more confusing? Up to you!

hud10837 commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,325 @@
+          // create a simple line symbol for the route
+          val routeSymbol = SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, Color.BLUE, 5f)
+          // create a graphic for the route
+          Graphic(route.routeGeometry, routeSymbol).also {
+            // add the graphic to the map
+            graphicsOverlay.graphics.add(it)
```
Would it be better to make this two lines? I'm happy for nesting, but this is a fairly complicated line 😅

hud10837 replied:
Looking at it in context, it's not so bad

hud10837 commented on kotlin/find-route/build.gradle:
```diff
@@ -0,0 +1,29 @@
+    // arcgis-android & appcompat lib dependencies from rootProject build.gradle
+    implementation "androidx.constraintlayout:constraintlayout:$constraintLayoutVersion"
+    // material design dependency for the bottom sheet behavior
+    implementation 'com.google.android.material:material:1.1.0'
```
it was on develop! This was a good reminder to rebase

hud10837 commented:
@TADraeseke I've got a few questions about your comments, but implemented the rest. I also rebased this off of develop because I believe we were discussing some potential bugs in master, and the material version parameter is only on develop right now. Finally, I modified the find-address sample because it wouldn't build; android studio was just asking for some double quotes instead of single quotes in the build.gradle.

TADraeseke commented on kotlin/find-route/src/main/res/layout/bottom_sheet.xml:
```diff
@@ -0,0 +1,44 @@
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="12dp"
+            android:layout_marginBottom="12dp"
+            android:text="Directions"
```
to string resource

TADraeseke commented on kotlin/find-route/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,55 @@
+        android:layout_height="match_parent"
+        android:layout_marginBottom="64dp">
+
+        <ProgressBar
```
I think this might need to go lower down than the map view in the xml file to fix a z-ordering issues where I don't see the bar

hud10837 replied:
wow, great catch!

TADraeseke commented on kotlin/find-route/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
```
unused

TADraeseke commented on kotlin/find-route/src/main/java/com/esri/arcgisruntime/sample/findroute/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+              SimpleLineSymbol.Style.SOLID,
+              Color.GREEN, 5f
+            )
+            Graphic(geometry, selectedRouteSymbol).also {
```
I'd just inline the graphic creation as above

hud10837 replied:
I agree, this was before we talked about this pattern!

hud10837 commented:
Thanks for getting to this @TADraeseke , I've made those changes and an additional change to fix a graphical error with the arrow on the bottom sheet. Can you take another look?

TADraeseke commented:
Thanks @hud10837 - Looking great! Just before you merge, can you run the linter (Analyze -> Inspect code... -> find route module). There's a couple unused resources and imports.

hud10837 commented:
@TADraeseke Wow, I should be doing that more often! I always thought the linter was just the code reformatter. I've changed those now
